Send only significant MAX_DATA frames (2)

Matthias247 opened the pull request:
This is an update to #893 which uses a newtype to indicate whether an outgoing frame needs to be enqueued.
The update is in the second commit.

#893 could be merged before this - or this can be extended and replace the orginal CR.

Matthias247 commented:
Based on implementing this, I found out that `#[must_use]` part definitely pays off and is useful. It should likely also be adopted to other frames which need to be enqueued.

Matthias247 commented on quinn-proto/src/connection/streams.rs:
```diff
@@ -172,17 +180,17 @@ impl Streams {
         id: StreamId,
         buf: &mut [u8],
-    ) -> Result<Option<(usize, bool)>, ReadError> {
+    ) -> Result<Option<(usize, bool, WantTransmit)>, ReadError> {
```
What I don't like here, is that if `MAX_STREAM_DATA` would change to the same thing, we would return `WantTransmit, WantTransmit`.

And thereby I get to what I don't think is perfect yet: The naming. 
Should it be a different newtype per frame-type? `WantTransmitMaxData` + `WantTransmitMaxStreamData`? At least makes it obvious in tuples what is what.

Another idea is making `WantTransmit` generic, and embedded the frame type, like `WantTransmit<MaxData>`. But that could be overengineered.

Another option is to simply use structs for these return values, instead of tuples.

Ralith replied:
3 items is my threshold for replacing structs with tuples, so I'm on board with that regardless. It might also be nice to disambiguate `WantTransmit`; a phantom type parameter using a frame type would be more pleasant than a bunch of functionally identical handwritten newtypes, for all that it's a bit cheeky. No strong opinion on whether that's really needed though.

Matthias247 replied:
I now added a dedicated result structs for all those operations

Ralith reviewed (commented):
Seems generally reasonable. Rebase needed.

djc commented:
I like the idea of a phantom type parameter for `WantTransmit`.

Matthias247 commented:
I tried the `PhantomType` approach, but it didn't work out. The reason is that there don't exist dedicated frame types for each those frames. Most are only enum variants within `frame::Frame`, and can't be used as type parameters. The pure `frame::Type::MAX_DATA` version also didn't work out - I guess it needs const-generics to make it work. Therefore this now has a single `WantTransmit` for everything.

However I updated a few more paths to make use of it.
In a few places I also replaced tuples with result structs to at least give the returned fields some names. I however feel like this has a negative impact regarding `#[must_use]`. I don't think `must_use` fields within structs are checked to be used, as long as the struct itself is used.

djc commented on quinn-proto/src/connection/streams.rs:
```diff
@@ -416,18 +429,23 @@ impl Streams {
     ///
-    /// Returns `true` iff the peer should be notified.
-    pub fn stop(&mut self, id: StreamId) -> Result<bool, UnknownStream> {
+    /// Returns structure which indicates whether this requires transmissing
```
`transmissing` is not a word, I think, and "whether this requires transmitting and additional frames" also seems wrong? Maybe "and" -> "any"?

djc commented on quinn-proto/src/frame.rs:
```diff
@@ -826,6 +826,29 @@ impl Datagram {
     }
 }
 
+/// Indicates whether a frame of type `F` needs to be transmitted
```
This documentation seems outdated, type `F` is no longer there.

djc commented on quinn-proto/src/frame.rs:
```diff
@@ -826,6 +826,29 @@ impl Datagram {
 
+/// Indicates whether a frame of type `F` needs to be transmitted
+///
+/// This type is around `bool` which uses the `#[must_use]` attribute in order
```
This sentence reads a bit odd to me, maybe "This type wraps around `bool` and uses the ..."?

djc commented on quinn-proto/src/frame.rs:
```diff
@@ -826,6 +826,29 @@ impl Datagram {
+    want_transmit: bool,
+}
+
+impl WantTransmit {
+    /// Creates a new `WantTransmit` instance, which indicates whether a frame
+    /// of type `F` should be transmitted
+    pub fn new(want_transmit: bool) -> Self {
+        Self { want_transmit }
+    }
+
+    /// Returns whether a frame ot type `F` should be transmitted
+    pub fn want_transmit(self) -> bool {
+        self.want_transmit
+    }
+}
```
These methods don't feel like they carrry their weight at this point (perhaps as an effect of losing the type argument?).

Ralith replied:
Maybe a one-public-element tuple struct? It's not like this can conceivably grow any internal logic in the future.

Or maybe an `enum`? That's a little more idiomatic, I think, but I'm not sure what good variant names would be.

Matthias247 replied:
If you don't have them, you end up with checks which look like `if transmit_max_data.0 {`

I'm personally not a fan of those `.0` accesses. But I agree the duplication between method and type names isn't elegant. If someone has ideas around better names, I'm all open for it. It could be as simple as `fn yes(&self)`, but that looks weird too.

Ralith replied:
Maybe `if max_data.transmit()`?

Matthias247 replied:
That looks good. But it's kind of debatable whether to name the field in `ReadResult` `max_data`, since it doesn't contain the `max_data` value. If we get to `result.should_transmit_max_data.should_transmit()` we have the doubling again.`  `result.transmit_max_data.transmit()` is kind of the same

djc replied:
I think `result.max_data.should_transmit()` reads exactly right -- it doesn't seem like a big concern that the `bool`-like value of `result.max_data` is not a `max_data` value, since that should be obvious from the type system.

djc commented on quinn-proto/src/connection/mod.rs:
```diff
@@ -2820,12 +2822,17 @@ where
+    fn add_read_credits(
+        &mut self,
+        id: StreamId,
+        transmit_max_stream_data: frame::WantTransmit,
```
Let's just import `WantTransmit` from `frame` directly.

Also, I feel like `ShouldTransmit` better matches the semantics than `WantTransmit`, do others feel the same?

Ralith replied:
I'm weakly in favor.

djc reviewed (commented):
Looks useful, but maybe a bit overcomplicated as is?

Ralith commented:
> there don't exist dedicated frame types for each those frames

I wouldn't mind having a type for every frame for consistency, but I think that's getting pretty scope-creepy for this PR.

> I don't think must_use fields within structs are checked to be used, as long as the struct itself is used.

This is an issue we've hit before, but I'm pretty sure it's true for tuples as well.

Matthias247 commented:
Clippy error is unrelated to these changes. I guess it's a 1.48 release artifact

djc commented:
> Clippy error is unrelated to these changes. I guess it's a 1.48 release artifact

Addressed those in #923.

djc commented on quinn-proto/src/connection/streams.rs:
```diff
@@ -288,21 +295,27 @@ impl Streams {
+        // We don't buffer data on stopped streams, so issue flow control credit immediately
+        let want_transmit = self.add_read_credits(new_bytes);
+
+        Ok(want_transmit)
```
Nit: let's inline the call to `self.add_read_credits()` in `Ok()`?

djc commented on quinn-proto/src/connection/streams.rs:
```diff
@@ -416,18 +426,23 @@ impl Streams {
             None => return Err(UnknownStream { _private: () }),
         };
         stream.assembler.stop();
+        let transmit_stop_sending = ShouldTransmit::new(!stream.is_finished());
```
Nit: let's call this `stop_sending` and `transmit_max_data` -> `max_data` so that the initialization in the return value is more streamlined.

djc reviewed (commented):
This is looking great, almost done!

djc commented:
Thanks!
